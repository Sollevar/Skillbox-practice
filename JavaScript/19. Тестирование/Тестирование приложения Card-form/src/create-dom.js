import { el, setChildren } from 'redom';
import './styles/style.css';
import mainImage from './assets/image.jpg';
export function createCardForm() {
  const circlesWrapper = el('div.circles');
  const circleOne = el('div.circle circle-1');
  const circleSecond = el('div.circle circle-2');
  const cardWrapper = el('div.card-group');
  const cardForm = el('form.card');
  const cvcSpan = el('span', 'CVC:');
  const cvcLabel = el('label.cvc');
  const cvcInput = el('input', { type: 'text' });
  const logo = el('div.logo', 'Type');
  const imgWrapper = el('div.chip');
  const img = el('img');
  const cardNumberInput = el('input.number', {
    type: 'text',
  });
  cardNumberInput.placeholder = 'Enter card number';
  const cardEmailInput = el('input.email', {
    type: 'email',
  });
  cardEmailInput.placeholder = 'Enter email';
  const cardDateInput = el('input.to', {
    type: 'text',
  });
  cvcInput.placeholder = 'enter cvc';
  const sendButton = el('button.btn-send', 'Pay');

  img.src = 'https://raw.githubusercontent.com/dasShounak/freeUseImages/main/chip.png';
  img.alt = 'chip';
  sendButton.style.backgroundImage = `url(${mainImage})`;
  sendButton.disabled = true;
  setChildren(cvcLabel, cvcSpan, cvcInput);
  setChildren(imgWrapper, img);
  setChildren(cardForm, cvcLabel, logo, imgWrapper, cardNumberInput, cardEmailInput, cardDateInput);
  setChildren(cardWrapper, cardForm, sendButton);
  setChildren(circlesWrapper, circleOne, circleSecond);
  setChildren(document.body, circlesWrapper, cardWrapper);
  return {
    cardNumberInput,
    cvcInput,
    cardDateInput,
    cardEmailInput,
    cardForm,
    sendButton,
    logo,
  };
}
