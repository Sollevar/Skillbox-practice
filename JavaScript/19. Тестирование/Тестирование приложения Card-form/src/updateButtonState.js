export const validityMap = new Map();
export function updateButtonState(sendButton) {
  sendButton.disabled = true;
  if (validityMap.size === 4) {
    const allValid = Array.from(validityMap.values()).every((valid) => valid);
    sendButton.disabled = !allValid;
  }
  return;
}
