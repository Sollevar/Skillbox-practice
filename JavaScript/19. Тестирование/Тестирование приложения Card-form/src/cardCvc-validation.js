import valid from 'card-validator';
import { updateButtonState, validityMap } from './updateButtonState.js';
export function cardCvcValidation(cvc, sendButton) {
  cvc.addEventListener('blur', () => {
    const value = valid.cvv(cvc.value);
    if (!value.isValid) {
      cvc.classList.add('not-valid');
      validityMap.set('cvc', false);
    } else {
      cvc.classList.remove('not-valid');
      validityMap.set('cvc', true);
    }
    updateButtonState(sendButton);
  });

  cvc.addEventListener('focus', () => {
    cvc.classList.remove('not-valid');
  });
}
