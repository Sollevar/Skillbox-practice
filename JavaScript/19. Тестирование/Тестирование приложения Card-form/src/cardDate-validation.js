import valid from 'card-validator';
import { updateButtonState, validityMap } from './updateButtonState.js';
export function cardDateValidation(date, sendButton) {
  date.addEventListener('blur', () => {
    const value = valid.expirationDate(date.value, 4);
    if (!value.isValid) {
      date.classList.add('not-valid');
      validityMap.set('date', false);
    } else {
      date.classList.remove('not-valid');
      validityMap.set('date', true);
    }
    updateButtonState(sendButton);
  });

  date.addEventListener('focus', () => {
    date.classList.remove('not-valid');
  });
}
