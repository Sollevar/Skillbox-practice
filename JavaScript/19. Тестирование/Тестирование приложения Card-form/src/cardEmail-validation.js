import { updateButtonState, validityMap } from './updateButtonState.js';
export function cardEmailValidation(email, sendButton) {
  email.addEventListener('blur', () => {
    const value = email.value.trim();
    if (!value.includes('@') || !value.includes('.')) {
      email.classList.add('not-valid');
      validityMap.set('email', false);
    } else {
      email.classList.remove('not-valid');
      validityMap.set('email', true);
    }
    updateButtonState(sendButton);
  });

  email.addEventListener('focus', () => {
    email.classList.remove('not-valid');
  });
}
