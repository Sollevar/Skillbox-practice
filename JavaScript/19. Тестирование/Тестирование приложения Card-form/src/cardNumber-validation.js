import valid from 'card-validator';
import { updateButtonState, validityMap } from './updateButtonState.js';
export function cardNumberValidation(number, logo, sendButton) {
  number.addEventListener('blur', () => {
    const value = valid.number(number.value.replace(/\s+/g, ''));
    if (!value.isValid) {
      number.classList.add('not-valid');
      logo.textContent = 'Type';
      validityMap.set('number', false);
    } else {
      number.classList.remove('not-valid');
      logo.textContent = value.card.niceType;
      validityMap.set('number', true);
    }
    updateButtonState(sendButton);
  });

  number.addEventListener('focus', () => {
    number.classList.remove('not-valid');
  });
}
