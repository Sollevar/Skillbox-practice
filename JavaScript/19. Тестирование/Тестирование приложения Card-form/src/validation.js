import Inputmask from '../node_modules/inputmask/dist/inputmask.es6.js';
import { cardNumberValidation } from './cardNumber-validation.js';
import { cardDateValidation } from './cardDate-validation.js';
import { cardCvcValidation } from './cardCvc-validation.js';
import { cardEmailValidation } from './cardEmail-validation.js';
import { updateButtonState, validityMap } from './updateButtonState.js';

export function validator(cardNumber, cardCvc, cardEmail, cardDate, cardForm, sendButton, logo) {
  const numberMask = new Inputmask('9999 9999 9999 9999');
  const cvcMask = new Inputmask('999');
  const dateMask = new Inputmask('99/99');
  numberMask.mask(cardNumber);
  cvcMask.mask(cardCvc);
  dateMask.mask(cardDate);

  cardNumberValidation(cardNumber, logo, sendButton);
  cardDateValidation(cardDate, sendButton);
  cardCvcValidation(cardCvc, sendButton);
  cardEmailValidation(cardEmail, sendButton);

  sendButton.addEventListener('click', () => {
    alert('Оплата прошла успешно');
    for (const input of cardForm.elements) {
      input.value = '';
    }
    validityMap.clear();
    updateButtonState(sendButton);
  });
}
