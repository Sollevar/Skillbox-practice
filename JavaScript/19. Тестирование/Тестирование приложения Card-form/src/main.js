import 'babel-polyfill';
import { createCardForm } from './create-dom.js';
import { validator } from './validation.js';
const { cardNumberInput, cvcInput, cardEmailInput, cardDateInput, cardForm, sendButton, logo } = createCardForm();
validator(cardNumberInput, cvcInput, cardEmailInput, cardDateInput, cardForm, sendButton, logo);
