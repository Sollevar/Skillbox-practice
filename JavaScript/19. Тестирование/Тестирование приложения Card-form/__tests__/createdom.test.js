import { createCardForm } from '../src/create-dom';

test('Создание формы банковской карты', () => {
  createCardForm();
  expect(document.body).toMatchSnapshot();
});
