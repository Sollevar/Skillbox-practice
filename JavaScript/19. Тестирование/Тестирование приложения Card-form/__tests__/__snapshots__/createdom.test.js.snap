// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Создание формы банковской карты 1`] = `
<body>
  <div
    class="circles"
  >
    <div
      class="circle circle-1"
    />
    <div
      class="circle circle-2"
    />
  </div>
  <div
    class="card-group"
  >
    <form
      class="card"
    >
      <label
        class="cvc"
      >
        <span>
          CVC:
        </span>
        <input
          placeholder="enter cvc"
          type="text"
        />
      </label>
      <div
        class="logo"
      >
        Type
      </div>
      <div
        class="chip"
      >
        <img
          alt="chip"
          src="https://raw.githubusercontent.com/dasShounak/freeUseImages/main/chip.png"
        />
      </div>
      <input
        class="number"
        placeholder="Enter card number"
        type="text"
      />
      <input
        class="email"
        placeholder="Enter email"
        type="email"
      />
      <input
        class="to"
        type="text"
      />
    </form>
    <button
      class="btn-send"
      disabled=""
      style="background-image: url(mocked-image-path);"
    >
      Pay
    </button>
  </div>
</body>
`;
